### ISSUE TITLE
allow internal parallelization

### ISSUE THREAD

**Issue by basnijholt** [pipefunc/pipefunc#191]

**Comment by basnijholt** [comment]
implemented a while ago

**Comment by bra-fsn** [comment]
> implemented a while ago

You mean the parallel `.map` function?

I thought this issue is about something like the following:
```python
import time
import threading
from pipefunc import pipefunc, Pipeline

@pipefunc(output_name="a")
def f_a():
    time.sleep(2)
    print("f_a", threading.current_thread(), time.time())

    return "a"

@pipefunc(output_name="b")
def f_b():
    time.sleep(2)
    print("f_b", threading.current_thread(), time.time())

    return "b"

@pipefunc(output_name="result")
def f_res(a, b):
    return a + b

pipeline = Pipeline([f_a, f_b, f_res])

print(pipeline("result"))
```
Which should run `f_a` and `f_b` in parallel (either in threads or processes), instead of sequentially.
Is this implemented?

**Comment by basnijholt** [comment]
Yes this is implemented. Although when calling the pipeline as in your example it will always be sequential. Using map will correctly parallelize everything. See https://pipefunc.readthedocs.io/en/latest/tutorial/#custom-parallelism

This issue, however, is related to setting resource requirements per function and then letting the function itself handle the parallelism.

**Comment by basnijholt** [comment]
For completeness, your example, when calling:
```
print(pipeline.map(inputs={}))
```
gives:
```
f_bf_a  <_MainThread(MainThread, started 8630751040)><_MainThread(MainThread, started 8630751040)>  1725486554.0917828
1725486554.091784
OrderedDict([('a', Result(function='f_a', kwargs={}, output_name='a', output='a', store=None, run_folder=PosixPath('/tmp/tmp5oldmzvp'))), ('b', Result(function='f_b', kwargs={}, output_name='b', output='b', store=None, run_folder=PosixPath('/tmp/tmp5oldmzvp'))), ('result', Result(function='f_res', kwargs={'a': 'a', 'b': 'b'}, output_name='result', output='ab', store=None, run_folder=PosixPath('/tmp/tmp5oldmzvp')))])
```

**Comment by bra-fsn** [comment]
> For completeness, your example, when calling:
> 
> ```
> print(pipeline.map(inputs={}))
> ```
> 
> gives:
> 
> ```
> f_bf_a  <_MainThread(MainThread, started 8630751040)><_MainThread(MainThread, started 8630751040)>  1725486554.0917828
> 1725486554.091784
> OrderedDict([('a', Result(function='f_a', kwargs={}, output_name='a', output='a', store=None, run_folder=PosixPath('/tmp/tmp5oldmzvp'))), ('b', Result(function='f_b', kwargs={}, output_name='b', output='b', store=None, run_folder=PosixPath('/tmp/tmp5oldmzvp'))), ('result', Result(function='f_res', kwargs={'a': 'a', 'b': 'b'}, output_name='result', output='ab', store=None, run_folder=PosixPath('/tmp/tmp5oldmzvp')))])
> ```

Very cool, thank you! I've misunderstood the "letting the function itself handle the parallelism".

One more question for the above example -I hope you don't mind if I ask it here-: as all methods are decorated with the output names they are responsible for, pipefunc in theory could discover the relevant providers for a given output, so if it requires that all output names are unique, something like this should be enough:

```python
pipeline = Pipeline(entry_func=f_res)
```
instead of listing all nodes in the graph.
Of course, there are some questions regarding this, like:
1. how deep it should look (like in imported modules' namespaces)
2. if it requires output_name uniqueness, the method name itself could work as well for naming/referencing the outputs (like in pytest's fixtures). Imported modules create some complexity here...

So for the same namespace, this could work:

```python
@pipefunc
def a():
    return "a"

@pipefunc
def b():
    return "b"

@pipefunc
def res(a, b):
    return a + b

pipeline = Pipeline(entry=res)
```

This won't allow to reference a method directly from a different module though, but that could be solved with annotations, like this:
```python
from pipefunc import pipefunc, Pipeline, Input
from typing import Annotated
import another_module

# in another_module:
# @pipefunc
# def a():
#     return "a"

@pipefunc
def b():
    return "b"

@pipefunc
def res(a: Annotated[str, Input("another_module.a")], b):
    return a, b

# this is the equivalent code without annotations:
# another_a = another_module.a
# @pipefunc
# def res(another_a, b):
#     return another_a, b


pipeline = Pipeline(entry=res)
```
Calling `Pipeline` it could discover all the graph's methods and then do the same it does today.
This way, you don't have to name outputs (as they are already named by their method names) and list all the graph nodes as they are implicitly defined by referencing the output(method) names.

What do you think?

**Comment by basnijholt** [comment]
Part of what you are asking for already exists:

```python
def a():
    return "a"

def b():
    return "b"

def res(a, b):
    return a + b

pipeline = Pipeline(entry=[a, b, res])
```

Note that if functions are not decorated with `@pipefunc`, it will use the function name as parameter name.

However, one still needs to be explicit about which functions go into the pipeline.

I am not sure about the auto-detection of functions and automatically putting them in a pipeline. I think there will be too many surprises.

```bash
at 10:27:30 ❯ python -m this | grep Explicit
Explicit is better than implicit.
```

😄 

However, if there are good use-cases you (or others) might be able to convince me!

**Comment by bra-fsn** [comment]
> Part of what you are asking for already exists:
> 
> ```python
> def a():
>     return "a"
> 
> def b():
>     return "b"
> 
> def res(a, b):
>     return a + b
> 
> pipeline = Pipeline(entry=[a, b, res])
> ```
> 
> Note that if functions are not decorated with `@pipefunc`, it will use the function name as parameter name.
Indeed, this works. Great!

> 
> However, one still needs to be explicit about which functions go into the pipeline.
> 
> I am not sure about the auto-detection of functions and automatically putting them in a pipeline. I think there will be too many surprises.

Will be?

I mean what's the difference between:
```python
def a():
    return "a"

def b():
    return "b"

def res(a, b):
    return a + b

pipeline = Pipeline([a, b, res])
```

and
```python
def a():
    return "a"

def b():
    return "b"

def res(a, b):
    return a + b

pipeline = Pipeline(res)
```
?
I'm not sure though whether it should be
`pipeline = Pipeline(res)`
or
`pipeline = Pipeline([res])`
or
`pipeline = Pipeline(entry=res)` (a new option, so it won't interfere with the current API)

In the former example, you have to explicitly reference `a` and `b` as method arguments, that's how they're looked up. Listing all methods in the graph is redundant, as you already had to build those edges by using method names as argument names.
I think that's pretty explicit.
Could you please elaborate on what surprises do you expect if you don't have to manually list all graph nodes, but simply use the referenced methods, like -thanks for the info- it works today and pass only the entry node to `Pipeline`?

Anyways, this is much cooler than I thought, thanks for creating this package!

**Comment by basnijholt** [comment]
I just typically know I will define many small helper functions that I don't want in the `Pipeline`. I could of course make some naming convention, like ignore `_functions` with an underscore.

I will accept a PR however I won't do this myself soon I think.

**Comment by basnijholt** [comment]
> Anyways, this is much cooler than I thought, thanks for creating this package!

@bra-fsn, thanks! I hope you'll try it!
